fullstack-cert/typescript-projects/flashcard-quiz-app/src/script.ts

```diff
@@ -0,0 +1,226 @@
+
+  try {
+    if (!frontText)
+      throw new InvalidUserInputError("Front text cannot be empty.");
+    if (!backText)
+      throw new InvalidUserInputError("Back text cannot be empty.");
+
+    const newCard: FlashCard = { frontText, backText };
+    currentCards.push(newCard);
+    const newIndex = currentCards.length - 1;
+    const cardBtn = createCardButton(frontText, newIndex);
+    cardButtonsContainer.appendChild(cardBtn);
+
+    currentCardIndex = newIndex;
+    showingFront = true;
+    refresh();
+
+    frontInput.value = "";
+    backInput.value = "";
+  } catch (ex) {
+    if (ex instanceof InvalidUserInputError) {
+      errorElement.textContent = "⚠️ " + ex.message;
+    } else {
+      console.error("An unexpected error occurred:", ex);
+    }
+  }
+}
+
+document.addEventListener("keydown", (ev: KeyboardEvent) => {
+  const frontInput = document.getElementById(
+    "front-text",
+  ) as HTMLTextAreaElement;
+  const backInput = document.getElementById("back-text") as HTMLTextAreaElement;
+  if (
+    ev.key === "Enter" &&
+    document.activeElement !== frontInput &&
+    document.activeElement !== backInput
+  ) {
+    ev.preventDefault();
+    flipCard();
+  }
+});
+
+interface Flashcard {
+  question: string;
+  answer: string;
+  isAnswered?: boolean;
+  isCorrect?: boolean;
+}
+
+interface CardDeck {
+  name: string;
+  cards: Flashcard[];
+}
+
+interface GameState {
+  currentCardIndex: number;
+  isFlipped: boolean;
+  correctCount: number;
+  wrongCount: number;
+  gameStarted: boolean;
+  gameCompleted: boolean;
+  currentDeck: Flashcard[];
+}
+
+const cardDecks: Record<string, CardDeck> = {
+  general: {
+    name: "General Knowledge",
+    cards: [
+      { question: "What is the capital of France?", answer: "Paris" },
+      { question: "Which planet is known as the Red Planet?", answer: "Mars" },
+      {
+        question: "What is the largest mammal in the world?",
+        answer: "Blue whale",
+      },
+      { question: "In which year did the Titanic sink?", answer: "1912" },
+    ],
+  },
+};
+
+// better to add a game class to encapsulate the game logic
+// This will help in managing the game state, current card, and user interactions.
+// This class can also handle the game flow, such as starting a new game, flipping cards
+// and checking answers, making the code more organized and maintainable.
+class FlashcardGame {
+  state: GameState;
+  private elements: {
+    flashcard: HTMLElement;
+    cardsList: HTMLElement;
+    flipBtn: HTMLElement;
+  };
+
+  constructor() {
+    this.state = {
+      currentCardIndex: 0,
+      isFlipped: false,
+      correctCount: 0,
+      wrongCount: 0,
+      gameStarted: false,
+      gameCompleted: false,
+      currentDeck: [...cardDecks.general.cards],
+    };
+
+    this.elements = {
+      flashcard: document.querySelector(".flashcard")!,
+      cardsList: document.getElementById("cards-list")!,
+      flipBtn: document.getElementById("flip-btn")!,
```

Comment:
Rather than using non-null assertions here, we should have the campers implement a fallback approach if the elements are not found.

Maybe something like injecting an element with text "DOM failed to hydrate, refresh the page"?